### PR TITLE
fix sklearn raw model extraction

### PR DESCRIPTION
**Pull request by serena-ruan**
## Related Issues/PRs

Fix

## What changes are proposed in this pull request?
If model_impl is not a _SklearnModelWrapper we directly return it.

## How is this patch tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/models/evaluation/default_evaluator.py
Comment (harupy) [inline mlflow/models/evaluation/default_evaluator.py:85]: Can we just return 
```suggestion
        return model_loader_module, model._model_impl
```

Doesn't this work?

Comment (serena-ruan) [reply mlflow/models/evaluation/default_evaluator.py:85]: If user wants to call some other functions on the raw model, like other than 'predict, predict_prob, or score', functions we didn't wrap in _SklearnModelWrapper, then this will be a problem

Comment (harupy) [reply mlflow/models/evaluation/default_evaluator.py:85]: `_model_impl` is a private field, we don't expect users to touch it.

File: tests/evaluate/test_default_evaluator.py
Comment (harupy) [inline tests/evaluate/test_default_evaluator.py:2226]: Instead of testing, can we test `mlflow.evaluate`?

Comment (harupy) [reply tests/evaluate/test_default_evaluator.py:2226]: Just because `_extract_raw_model` works fine doesn't mean it works fine in `mlflow.evaluate.`

Comment (harupy) [reply tests/evaluate/test_default_evaluator.py:2226]: In general, we should test a public API, instead of implementatio details like `_extract_raw_model`.

**Review by harupy — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/9050/files#r1260759295 is addressed!

**Comment by serena-ruan** [comment]
@harupy https://github.com/mlflow/mlflow/blob/797cdf5cbbd238338b61f008b9308b246a5f4d4f/mlflow/models/evaluation/default_evaluator.py#L722-L732 I think this part depends on raw_model to be a sklearn model instead of the wrapper? If we return the wrapper, `evaluate` doesn't log those images. Should we return .sklearn_model or update the logic in default_evaluator to be compatible with the wrapper instead?

**Comment by harupy** [comment]
Does `shap.Explainer` reject the wrapper instance?

**Comment by serena-ruan** [comment]
> Does `shap.Explainer` reject the wrapper instance?

I think so. I added an Exception here:
image

It returns directly, so these three images are not logged.

File: tests/evaluate/test_default_evaluator.py
Comment (harupy) [inline tests/evaluate/test_default_evaluator.py:2234]: Do we need this?

File: mlflow/sklearn/__init__.py
Comment (harupy) [inline mlflow/sklearn/__init__.py:517]: Do we need this condition?

File: tests/evaluate/test_default_evaluator.py
Comment (serena-ruan) [reply tests/evaluate/test_default_evaluator.py:2234]: Yes, otherwise there's ` AttributeError: 'SubRequest' object has no attribute 'param'` error

File: mlflow/sklearn/__init__.py
Comment (harupy) [reply mlflow/sklearn/__init__.py:517]: This means this function returns None, which can lead to a hard-to-debug error. Just throwing `sklearn_model has no attribute predict_proba` might easier to debug.

Comment (serena-ruan) [reply mlflow/sklearn/__init__.py:517]: Yes, for this test: https://github.com/mlflow/mlflow/actions/runs/5527963781/jobs/10084241551#step:8:3032
on commit https://github.com/mlflow/mlflow/pull/9050/commits/e06bf73859a6682a4d39db41d452b08d207fa9ee
```
FAILED tests/evaluate/test_default_evaluator.py::test_svm_classifier_evaluation_disable_logging_metrics_and_artifacts - AttributeError: 'LinearSVC' object has no attribute 'predict_proba'
```

Comment (harupy) [reply mlflow/sklearn/__init__.py:517]: Interesting, how did this test pass before you made the inference params change?

File: tests/evaluate/test_default_evaluator.py
Comment (harupy) [reply tests/evaluate/test_default_evaluator.py:2234]: got it, let's keep this

File: tests/evaluate/test_default_evaluator.py
Comment (harupy) [inline tests/evaluate/test_default_evaluator.py:2227]: I don't think we need this test.

File: mlflow/sklearn/__init__.py
Comment (serena-ruan) [reply mlflow/sklearn/__init__.py:517]: Because `self.predict_fn, self.predict_proba_fn = _extract_predict_fn(model, self.raw_model)`. If the raw_model is the underlying sklearn_model, then from here it is the sklearn model's predict_proba function, instead of the wrapper's predict_proba. But if raw_model is a _SklearnModelWrapper, then `self.predict_prob_fn` always exist, we might need to update the logic here again since `predict_proba_fn = getattr(raw_model, "predict_proba", None)` is not None even the function itself returns None 🤔

Comment (serena-ruan) [reply mlflow/sklearn/__init__.py:517]: Or it's fine, because 
```
if self.predict_proba_fn is not None:
    self.y_probs = self.predict_proba_fn(self.X.copy_to_avoid_mutation())
else:
    self.y_probs = None
```
in the first situation it returns None as well

File: mlflow/models/evaluation/default_evaluator.py
Comment (WeichenXu123) [inline mlflow/models/evaluation/default_evaluator.py:85]: What's the case it is not `_SklearnModelWrapper` instance ? sklearn model._model_impl should be returned by sklearn flavor `_load_pyfunc`

Comment (serena-ruan) [reply mlflow/models/evaluation/default_evaluator.py:85]: There's a test case in universe/automl that directly construct a PyFuncModel by passing _model_impl which uses sklearn Pipeline model.

**Comment by serena-ruan** [comment]
Will not merge this PR to master, the failing test will be fixed automatically due to the revert PR: https://github.com/mlflow/mlflow/pull/9059

File: mlflow/models/evaluation/default_evaluator.py
Comment (WeichenXu123) [reply mlflow/models/evaluation/default_evaluator.py:85]: emm... is it a legal usage ?

Comment (WeichenXu123) [reply mlflow/models/evaluation/default_evaluator.py:85]: I guess other flavor might have similar issue if we directly construct `_model_impl`, but `_model_impl` is internal attribute managed by pyfunc if I understand it correctly

Comment (WeichenXu123) [reply mlflow/models/evaluation/default_evaluator.py:85]: I think if someone want to customize `_model_impl` attribute in sklearn pyfunc model,

then it should not be regarded as sklearn pyfunc model but he can create a custom pyfunc model that inherits `PythonModel` base class.

**Comment by serena-ruan** [comment]
As we discussed, in _extract_raw_model function, we still need to check https://github.com/mlflow/mlflow/blob/db6a0a6934e0fb433afde88710920f01de94243e/mlflow/models/evaluation/default_evaluator.py#L83-L84

We can refactor this part later if we wanna add a Base wrapper :)

File: mlflow/models/evaluation/default_evaluator.py
Comment (harupy) [inline mlflow/models/evaluation/default_evaluator.py:84]: Can we add a comment on why we need this?

Comment (harupy) [reply mlflow/models/evaluation/default_evaluator.py:84]: I'm also curious about what happens if we evaluate a non-sklearn model like xgboost.

Comment (serena-ruan) [reply mlflow/models/evaluation/default_evaluator.py:84]: I think there're two problems:
1. https://github.com/mlflow/mlflow/blob/4f76275317accb0b51770a4a83b8d1db2fda21f9/mlflow/models/evaluation/default_evaluator.py#L93-L103 The raw_model will never be xgboost.XGBModel as it will be the wrapper, and predict_prob_fn will be None for all wrappers as they're not implemented.
2. https://github.com/mlflow/mlflow/blob/4f76275317accb0b51770a4a83b8d1db2fda21f9/mlflow/models/evaluation/default_evaluator.py#L730-L732 Looks like `shap.Explainer` doesn't accept wrappers

**Pull request merged**
